flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/WindowGroupReorderRule.java

```diff
@@ -0,0 +1,192 @@
+                    } else {
+                        return keyComp;
+                    }
+                });
+
+        if (!sequenceGroups.equals(oldGroups) && !Lists.reverse(sequenceGroups).equals(oldGroups)) {
+            int offset = input.getRowType().getFieldCount();
+            List<int[]> aggTypeIndexes = new ArrayList<>();
+            for (Group group : oldGroups) {
+                int aggCount = group.aggCalls.size();
+                int[] typeIndexes = new int[aggCount];
+                for (int i = 0; i < aggCount; i++) {
+                    typeIndexes[i] = offset + i;
+                }
+                offset += aggCount;
+                aggTypeIndexes.add(typeIndexes);
+            }
+
+            offset = input.getRowType().getFieldCount();
+            List<Integer> mapToOldTypeIndexes =
+                    IntStream.range(0, offset).boxed().collect(Collectors.toList());
+            for (Group newGroup : sequenceGroups) {
+                int aggCount = newGroup.aggCalls.size();
+                int oldIndex = oldGroups.indexOf(newGroup);
+                offset += aggCount;
+                for (int aggIndex = 0; aggIndex < aggCount; aggIndex++) {
+                    mapToOldTypeIndexes.add(aggTypeIndexes.get(oldIndex)[aggIndex]);
+                }
+            }
+
+            List<Map.Entry<String, RelDataType>> newFieldList =
+                    mapToOldTypeIndexes.stream()
+                            .map(index -> window.getRowType().getFieldList().get(index))
+                            .collect(Collectors.toList());
+            RelDataType intermediateRowType =
+                    window.getCluster().getTypeFactory().createStructType(newFieldList);
+            LogicalWindow newLogicalWindow =
+                    LogicalWindow.create(
+                            window.getCluster().getPlanner().emptyTraitSet(),
+                            input,
+                            window.constants,
+                            intermediateRowType,
+                            sequenceGroups);
+
+            List<Integer> sortedIndices =
+                    IntStream.range(0, mapToOldTypeIndexes.size())
+                            .boxed()
+                            .sorted(Comparator.comparingInt(i -> mapToOldTypeIndexes.get(i)))
+                            .collect(Collectors.toList());
+
+            List<RexInputRef> projects =
+                    sortedIndices.stream()
+                            .map(
+                                    index ->
+                                            new RexInputRef(
+                                                    index, newFieldList.get(index).getValue()))
+                            .collect(Collectors.toList());
+
+            LogicalProject project =
+                    LogicalProject.create(
+                            newLogicalWindow,
+                            java.util.Collections.emptyList(),
```

Comment:
why can't we have `Collections.emptyList()` here?

Reply:
done @snuyanzin